crates/ruff/src/args.rs

```diff
@@ -594,12 +695,115 @@ fn resolve_bool_arg(yes: bool, no: bool) -> Option<bool> {
     }
 }
 
+/// `--config` arguments passed via the CLI.
+///
+/// Users may pass 0 or 1 paths to a configuration file,
+/// and an arbitrary number of "inline TOML" options.
+///
+/// For example:
+///
+/// ```sh
+/// ruff check --config "path/to/pyproject.toml" --config "max-line-length=90" --config "isort.case-sensitive=false"
+/// ```
+#[derive(Clone, Debug)]
+pub enum ConfigOption {
+    PathToConfigFile {
+        path: PathBuf,
+        arg: Option<String>,
+        cmd: clap::Command,
+    },
+    ConfigOverride(Arc<Options>),
+}
+
+#[derive(Clone)]
+pub struct ConfigOptionParser;
+
+impl ValueParserFactory for ConfigOption {
+    type Parser = ConfigOptionParser;
+
+    fn value_parser() -> Self::Parser {
+        ConfigOptionParser
+    }
+}
+
+impl TypedValueParser for ConfigOptionParser {
+    type Value = ConfigOption;
+
+    fn parse_ref(
+        &self,
+        cmd: &clap::Command,
+        arg: Option<&clap::Arg>,
+        value: &std::ffi::OsStr,
+    ) -> Result<Self::Value, clap::Error> {
+        let value = value
+            .to_str()
+            .ok_or_else(|| clap::Error::new(clap::error::ErrorKind::InvalidUtf8))?;
+        let path_to_config_file = PathBuf::from(value);
+        if path_to_config_file.exists() {
+            return Ok(ConfigOption::PathToConfigFile {
+                path: path_to_config_file,
+                arg: arg.map(clap::Arg::to_string),
+                cmd: cmd.clone(),
+            });
+        }
+        let toml_parse_error = match toml::from_str(value) {
+            Ok(option) => return Ok(ConfigOption::ConfigOverride(Arc::new(option))),
+            Err(toml_error) => toml_error,
+        };
+        let mut new_error = clap::Error::new(clap::error::ErrorKind::ValueValidation).with_cmd(cmd);
+        if let Some(arg) = arg {
+            new_error.insert(
```

Comment:
Nit: Some empty lines can help to make the early return easier to spot
```suggestion
        };
        
        let mut new_error = clap::Error::new(clap::error::ErrorKind::ValueValidation).with_cmd(cmd);
        if let Some(arg) = arg {
            new_error.insert(
```